osu.Game/Overlays/FirstRunSetup/ScreenWelcome.cs

```diff
@@ -20,13 +29,153 @@ private void load()
+                    {
+                        // Avoid height changes when changing language.
+                        new Dimension(GridSizeMode.AutoSize, minSize: 100),
+                    },
+                    Content = new[]
+                    {
+                        new Drawable[]
+                        {
+                            new OsuTextFlowContainer(cp => cp.Font = OsuFont.Default.With(size: CONTENT_FONT_SIZE))
+                            {
+                                Text = FirstRunSetupOverlayStrings.WelcomeDescription,
+                                RelativeSizeAxes = Axes.X,
+                                AutoSizeAxes = Axes.Y
+                            },
+                        },
+                    }
                 },
+                new LanguageSelectionFlow
+                {
+                    RelativeSizeAxes = Axes.X,
+                    AutoSizeAxes = Axes.Y
+                }
             };
         }
+
+        private class LanguageSelectionFlow : FillFlowContainer
+        {
+            private Bindable<string> frameworkLocale = null!;
+
+            [BackgroundDependencyLoader]
+            private void load(FrameworkConfigManager frameworkConfig)
+            {
+                Direction = FillDirection.Full;
+                Spacing = new Vector2(5);
+
+                ChildrenEnumerable = Enum.GetValues(typeof(Language))
+                                         .Cast<Language>()
+                                         .Select(l => new LanguageButton(l)
+                                         {
+                                             Action = () => frameworkLocale.Value = l.ToCultureCode()
+                                         });
+
+                frameworkLocale = frameworkConfig.GetBindable<string>(FrameworkSetting.Locale);
+                frameworkLocale.BindValueChanged(locale =>
+                {
+                    if (!LanguageExtensions.TryParseCultureCode(frameworkLocale.Value, out var language))
+                        language = Language.en;
+
+                    foreach (var c in Children.OfType<LanguageButton>())
+                        c.Selected = c.Language == language;
+                }, true);
+            }
+
+            private class LanguageButton : OsuClickableContainer
+            {
+                public readonly Language Language;
+
+                private Box backgroundBox = null!;
+
+                private OsuSpriteText text = null!;
+
+                [Resolved]
+                private OverlayColourProvider colourProvider { get; set; } = null!;
+
+                private bool selected;
+
+                public bool Selected
+                {
+                    get => selected;
+                    set
+                    {
+                        if (selected == value)
+                            return;
+
+                        selected = value;
+
+                        updateState();
```

Comment:
Not super sure this is very safe. `updateState()` adds transforms, so it is potentially susceptible to invalid thread mutation errors. The usual pattern of doing this would consist of checking `IsLoaded` here and running `updateState()` again in `LoadComplete()` to accommodate potential value changes during load - is there any reason not to do the same here?

Also as an aside, I'm relatively sure that the state update animations are running when the button is shown initially. Usually remedied with a `FinishTransforms()` call, not sure if intentional or not.

Reply:
The initial animation was intended. I'll protect against the case you mentioned (considered it but also checked and it didn't seem like it would occur under normal operations).